internal/cmd/vet.go

```diff
@@ -0,0 +1,189 @@
+package cmd
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"runtime/trace"
+	"strings"
+
+	"github.com/google/cel-go/cel"
+	"github.com/spf13/cobra"
+
+	"github.com/kyleconroy/sqlc/internal/config"
+	"github.com/kyleconroy/sqlc/internal/debug"
+	"github.com/kyleconroy/sqlc/internal/opts"
+	"github.com/kyleconroy/sqlc/internal/plugin"
+)
+
+var ErrFailedChecks = errors.New("failed checks")
+
+func NewCmdVet() *cobra.Command {
+	return &cobra.Command{
+		Use:   "vet",
+		Short: "Vet examines queries",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			defer trace.StartRegion(cmd.Context(), "vet").End()
+			stderr := cmd.ErrOrStderr()
+			dir, name := getConfigPath(stderr, cmd.Flag("file"))
+			if err := Vet(cmd.Context(), ParseEnv(cmd), dir, name, stderr); err != nil {
+				if !errors.Is(err, ErrFailedChecks) {
+					fmt.Fprintf(stderr, "%s\n", err)
+				}
+				os.Exit(1)
+			}
+			return nil
+		},
+	}
+}
+
+func Vet(ctx context.Context, e Env, dir, filename string, stderr io.Writer) error {
+	configPath, conf, err := readConfig(stderr, dir, filename)
+	if err != nil {
+		return err
+	}
+
+	base := filepath.Base(configPath)
+	if err := config.Validate(conf); err != nil {
+		fmt.Fprintf(stderr, "error validating %s: %s\n", base, err)
+		return err
+	}
+
+	if err := e.Validate(conf); err != nil {
+		fmt.Fprintf(stderr, "error validating %s: %s\n", base, err)
+		return err
+	}
+
+	env, err := cel.NewEnv(
+		cel.StdLib(),
+		cel.Types(
+			&plugin.VetConfig{},
+			&plugin.VetQuery{},
+		),
+		cel.Variable("query",
+			cel.ObjectType("plugin.VetQuery"),
+		),
+		cel.Variable("config",
+			cel.ObjectType("plugin.VetConfig"),
+		),
+	)
+	if err != nil {
+		return fmt.Errorf("new env; %s", err)
+	}
+
+	checks := map[string]cel.Program{}
+	msgs := map[string]string{}
+
+	for _, c := range conf.Rules {
+		if c.Name == "" {
```

Comment:
Did you consider putting some of this basic validation inside of a shared config parse/validation area? Not sure if config.Validate() would be the right spot, or somewhere else, or maybe this idea just isn't correct if you think each subcommand should be responsible for validating its own "section" of the config (if it has one)?